File: leetcode/arrays-hashing/group_anagrams.ts
function groupAnagrams(strs: string[]): string[][] {
  const grouped = new Map<string, string[]>();
  for (const word of strs) {
    const key = word.split('').sort().join('');
    if (!grouped.has(key)) grouped.set(key, []);
    grouped.get(key)!.push(word);
  }
  return Array.from(grouped.values());
}

// Try it out — press Run
console.log(groupAnagrams(["eat","tea","tan","ate","nat","bat"]));
// [['eat','tea','ate'], ['tan','nat'], ['bat']] (order may vary)
console.log(groupAnagrams([""]));  // [['']]
console.log(groupAnagrams(["a"])); // [['a']]

export {};

File: leetcode/arrays-hashing/group_anagrams.py
from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        grouped = {}
        for word in strs:
            sorted_key = "".join(sorted(word))
            if sorted_key not in grouped:
                grouped[sorted_key] = []
            grouped[sorted_key].append(word)
        return list(grouped.values())

# Try it out — press Run
sol = Solution()
print(sol.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))
# [['eat','tea','ate'], ['tan','nat'], ['bat']] (order may vary)
print(sol.groupAnagrams([""]))   # [['']]
print(sol.groupAnagrams(["a"]))  # [['a']]
